File: lista/GO/questao4/animal.go
package main

import (
	"fmt"
)

type Animal interface {
	Som() string
}

type Cachorro struct {
	Nome string
}

func (c Cachorro) Som() string {
	return "Au au"
}

type Gato struct {
	Nome string
}

func (g Gato) Som() string {
	return "Miau"
}

func main() {
	cachorro := Cachorro{Nome: "Rex"}
	gato := Gato{Nome: "Michonne"}

	animais := []Animal{cachorro, gato}

	for _, animal := range animais {
		fmt.Println(animal.Som())
	}
}

File: lista/JAVA/questao4/Animal.java
package questao4;
/*
Crie uma classe base Animal com métodos como som. Derive classes como
Cachorro e Gato que implementam o método som de maneiras diferentes.
 */
public class Animal {
    protected String especie;
    protected String nome;
    public Animal(String especie, String nome){
        this.especie = especie;
        this.nome = nome;
    }
    public String emitirSom(){
        return "";
    }
}
class Cachorro extends Animal{
    public Cachorro(String nome){
        super("Cachorro", nome);
    }
    @Override
    public String emitirSom(){
        return "Au au";
    }
}
class Gato extends Animal{
    public Gato(String nome){
        super("Gato", nome);
    }
    @Override
    public String emitirSom(){
        return "Miau";
    }
}

class main{
    public static void main(String[] args) {
        Animal cachorro = new Cachorro("Rex");
        Animal gato = new Gato("Michonne");

        System.out.println(cachorro.emitirSom());
        System.out.println(gato.emitirSom());
    }
}
